tests/cli/demultiplex/test_verify_syncing.py

```diff
@@ -0,0 +1,40 @@
+from pathlib import Path
+from unittest import mock
+
+from cg.cli.demultiplex.demux import create_manifest_files
+
+
+def test_create_manifest_files_true(tmp_path: Path, cli_runner):
+    # GIVEN two flowcell directories
+    flow_cell_directories = [Path(tmp_path, "flowcell1"), Path(tmp_path, "flowcell2")]
+    for directory in flow_cell_directories:
+        directory.mkdir()
+
+    # GIVEN that both flowcell directories need a manifest file
+    with mock.patch("cg.cli.demultiplex.demux.needs_manifest_file", return_value=True):
+        with mock.patch(
+            "cg.cli.demultiplex.demux.create_manifest_file", return_value=None
+        ) as manifest_creation:
+            # WHEN creating manifest files
+            cli_runner.invoke(cli=create_manifest_files, args=["-s", tmp_path.as_posix()])
+
+    # THEN the manifest files creation should be called for every flowcell directory
+    assert manifest_creation.call_count == len(flow_cell_directories)
+
+
+def test_create_manifest_files_false(tmp_path: Path, cli_runner):
+    # GIVEN two flowcell directories
+    flow_cell_directories = [Path(tmp_path, "flowcell1"), Path(tmp_path, "flowcell2")]
+    for directory in flow_cell_directories:
+        directory.mkdir()
+
+    # GIVEN that none of the flowcell directories need a manifest file
+    with mock.patch("cg.cli.demultiplex.demux.needs_manifest_file", return_value=False):
+        with mock.patch(
+            "cg.cli.demultiplex.demux.create_manifest_file", return_value=None
+        ) as manifest_creation:
+            # WHEN creating manifest files
+            cli_runner.invoke(cli=create_manifest_files, args=["-s", tmp_path.as_posix()])
+
+    # THEN the manifest files creation should not be called
+    assert not manifest_creation.called
```

Comment:
I felt like I was testing the same functionality over and over again and thought this might be a good way to avoid that. Basically the create_manifest_files CLI command iterates trough directories checks if manifest files are needed and if so calls the `create_manifest_file` function.

These two tests only test that logic and assumes that the underlying functions are tested and behave as they should. Do you think this is a good way of testing higher level functions?

Reply:
I'm not a big fan of mocking internal functions because it makes the tests really brittle. If `needs_manifest_file` or `create_manifest_file` are renamed or something about their implementation changes, these tests will break. If a lot of tests use this pattern, devs will likely hesitate before renaming stuff.

I'd prefer it if we had unit tests for the most important functions and then integration tests for the main flow, which does not mock away internal functions.

I think mocking should be limited to calls against external resources, like a database/api. If it is the file system that needs to be mocked in this instance, maybe that can be done in a fixture separate from the tests.

Reply:
👍

Reply:
Maybe this CLI command is not complex enough to merit an integration test and it is enough with the unit tests for the functions that it calls

Reply:
Very good points! But regarding this:

> I'd prefer it if we had unit tests for the most important functions and then integration tests (like these) for the main flow, which does not mock away internal functions and then verifies that the actual side effects occurred.

Does that not end up being; write complete tests for everything? I'm okay with that but it is often quite tedious.

Reply:
I think it looks good, but I understand @seallard 's points. Maybe the fixture solution could be applied here

Reply:
> Very good points! But regarding this:
> 
> > I'd prefer it if we had unit tests for the most important functions and then integration tests (like these) for the main flow, which does not mock away internal functions and then verifies that the actual side effects occurred.
> 
> Does that not end up being; write complete tests for everything? I'm okay with that but it is often quite tedious.

No, that is true. We should be strategical about what we chose to test and how

Reply:
I will try to compile some guidelines about testing and present them you you guys next week for feedback